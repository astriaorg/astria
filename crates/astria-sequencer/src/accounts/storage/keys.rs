use std::borrow::Cow;

use astria_core::primitive::v1::asset::IbcPrefixed;
use astria_eyre::eyre::{
    ContextCompat as _,
    Result,
    WrapErr as _,
};

use crate::{
    accounts::AddressBytes,
    storage::keys::{
        AccountPrefixer,
        Asset,
    },
};

pub(in crate::accounts) const TRANSFER_BASE_FEE: &str = "accounts/transfer_base_fee";
const COMPONENT_PREFIX: &str = "accounts/";
const BALANCE_PREFIX: &str = "balance/";
const NONCE: &str = "nonce";

/// Example: `accounts/gGhH....zZ4=/balance/`.
///                   |base64 chars|
pub(in crate::accounts) fn balance_prefix<TAddress: AddressBytes>(address: &TAddress) -> String {
    format!(
        "{}/{BALANCE_PREFIX}",
        AccountPrefixer::new(COMPONENT_PREFIX, address)
    )
}

/// Example: `accounts/gGhH....zZ4=/balance/0202....0202`.
///                   |base64 chars|       |64 hex chars|
pub(in crate::accounts) fn balance<'a, TAddress, TAsset>(
    address: &TAddress,
    asset: &'a TAsset,
) -> String
where
    TAddress: AddressBytes,
    &'a TAsset: Into<Cow<'a, IbcPrefixed>>,
{
    format!(
        "{}/{BALANCE_PREFIX}{}",
        AccountPrefixer::new(COMPONENT_PREFIX, address),
        Asset::from(asset)
    )
}

/// Example: `accounts/gGhH....zZ4=/nonce`.
///                   |base64 chars|
pub(in crate::accounts) fn nonce<TAddress: AddressBytes>(address: &TAddress) -> String {
    format!(
        "{}/{NONCE}",
        AccountPrefixer::new(COMPONENT_PREFIX, address)
    )
}

pub(in crate::accounts) fn extract_asset_from_key(key: &str) -> Result<IbcPrefixed> {
    Ok(key
        .strip_prefix(COMPONENT_PREFIX)
        .and_then(|s| s.split_once(BALANCE_PREFIX).map(|(_, asset)| asset))
        .wrap_err("failed to strip prefix from account balance key")?
        .parse::<Asset>()
        .wrap_err("failed to parse storage key suffix as address hunk")?
        .get())
}

#[cfg(test)]
mod tests {
    use astria_core::primitive::v1::{
        asset::Denom,
        Address,
    };

    use super::*;

    fn address() -> Address {
        "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
            .parse()
            .unwrap()
    }

    fn asset() -> Denom {
        "an/asset/with/a/prefix".parse().unwrap()
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(TRANSFER_BASE_FEE);
        insta::assert_snapshot!(balance(&address(), &asset()));
        insta::assert_snapshot!(nonce(&address()));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(TRANSFER_BASE_FEE.starts_with(COMPONENT_PREFIX));
        assert!(balance(&address(), &asset()).starts_with(COMPONENT_PREFIX));
        assert!(nonce(&address()).starts_with(COMPONENT_PREFIX));
    }

    #[test]
    fn balance_prefix_should_be_prefix_of_balance_key() {
        assert!(balance(&address(), &asset()).starts_with(&balance_prefix(&address())));
    }

    #[test]
    fn should_extract_asset_from_key() {
        let asset = IbcPrefixed::new([2; 32]);
        let key = balance(&[1; 20], &asset);
        let recovered_asset = extract_asset_from_key(&key).unwrap();
        assert_eq!(asset, recovered_asset);
    }
}
