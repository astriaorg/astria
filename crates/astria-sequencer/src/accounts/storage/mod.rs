pub(super) mod keys;
mod values;

pub(crate) use values::Value;
pub(super) use values::{
    Balance,
    Nonce,
};
