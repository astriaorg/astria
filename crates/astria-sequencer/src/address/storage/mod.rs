pub(super) mod keys;
mod values;

pub(super) use values::AddressPrefix;
pub(crate) use values::Value;
