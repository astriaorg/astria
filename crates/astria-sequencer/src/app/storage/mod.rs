mod values;

pub(crate) use values::{
    BlockHeight,
    BlockTimestamp,
    ChainId,
    RevisionNumber,
    StorageVersion,
    Value,
};
