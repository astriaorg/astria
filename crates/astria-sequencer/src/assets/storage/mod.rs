pub(super) mod keys;
mod values;

pub(super) use values::TracePrefixedDenom;
pub(crate) use values::Value;
