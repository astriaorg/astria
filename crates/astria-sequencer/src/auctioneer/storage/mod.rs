pub(super) mod keys;
mod values;

pub(super) use values::EnshrinedAuctioneerEntry;
pub(crate) use values::Value;
