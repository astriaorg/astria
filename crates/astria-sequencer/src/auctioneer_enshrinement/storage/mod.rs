pub(super) mod keys;
mod values;