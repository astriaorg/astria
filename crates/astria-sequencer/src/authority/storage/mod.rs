mod values;

pub(crate) use values::{
    AddressBytes,
    ValidatorSet,
    Value,
};
