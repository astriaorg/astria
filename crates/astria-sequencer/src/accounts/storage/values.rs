use astria_eyre::eyre::bail;
use borsh::{
    BorshDeserialize,
    BorshSerialize,
};

#[derive(Debug, BorshSerialize, BorshDeserialize)]
pub(crate) struct Value(ValueImpl);

#[derive(Debug, BorshSerialize, BorshDeserialize)]
enum ValueImpl {
    Balance(Balance),
    Nonce(Nonce),
}

#[derive(Debug, BorshSerialize, BorshDeserialize)]
pub(in crate::accounts) struct Balance(u128);

impl From<u128> for Balance {
    fn from(balance: u128) -> Self {
        Balance(balance)
    }
}

impl From<Balance> for u128 {
    fn from(balance: Balance) -> Self {
        balance.0
    }
}

impl From<Balance> for crate::storage::StoredValue<'_> {
    fn from(balance: Balance) -> Self {
        crate::storage::StoredValue::Accounts(Value(ValueImpl::Balance(balance)))
    }
}

impl<'a> TryFrom<crate::storage::StoredValue<'a>> for Balance {
    type Error = astria_eyre::eyre::Error;

    fn try_from(value: crate::storage::StoredValue<'a>) -> Result<Self, Self::Error> {
        let crate::storage::StoredValue::Accounts(Value(ValueImpl::Balance(balance))) = value
        else {
            bail!("accounts stored value type mismatch: expected balance, found {value:?}");
        };
        Ok(balance)
    }
}

#[derive(Debug, BorshSerialize, BorshDeserialize)]
pub(in crate::accounts) struct Nonce(u32);

impl From<u32> for Nonce {
    fn from(nonce: u32) -> Self {
        Nonce(nonce)
    }
}

impl From<Nonce> for u32 {
    fn from(nonce: Nonce) -> Self {
        nonce.0
    }
}

impl From<Nonce> for crate::storage::StoredValue<'_> {
    fn from(nonce: Nonce) -> Self {
        crate::storage::StoredValue::Accounts(Value(ValueImpl::Nonce(nonce)))
    }
}

impl<'a> TryFrom<crate::storage::StoredValue<'a>> for Nonce {
    type Error = astria_eyre::eyre::Error;

    fn try_from(value: crate::storage::StoredValue<'a>) -> Result<Self, Self::Error> {
        let crate::storage::StoredValue::Accounts(Value(ValueImpl::Nonce(nonce))) = value else {
            bail!("accounts stored value type mismatch: expected nonce, found {value:?}");
        };
        Ok(nonce)
    }
}

#[cfg(test)]
mod tests {
    use std::mem::discriminant;

    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn value_impl_balance_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_balance_discriminant",
            format!("{:?}", discriminant(&ValueImpl::Balance(Balance(0))))
        );
    }

    #[test]
    fn value_impl_nonce_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_nonce_discriminant",
            format!("{:?}", discriminant(&ValueImpl::Nonce(Nonce(0))))
        );
    }

    // Note: This test must be here instead of in `crate::storage` since `ValueImpl` is not
    // re-exported.
    #[test]
    fn stored_value_account_discriminant_unchanged() {
        use crate::storage::StoredValue;
        assert_snapshot!(
            "stored_value_account_discriminant",
            format!(
                "{:?}",
                discriminant(&StoredValue::Accounts(Value(ValueImpl::Balance(Balance(
                    0
                )))))
            )
        );
    }
}
