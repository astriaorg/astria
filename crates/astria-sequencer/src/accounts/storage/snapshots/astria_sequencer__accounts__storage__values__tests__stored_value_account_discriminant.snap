---
source: crates/astria-sequencer/src/accounts/storage/values.rs
expression: "format!(\"{:?}\",\ndiscriminant(&StoredValue::Accounts(Value(ValueImpl::Balance(Balance(0))))))"
---
Discriminant(3)
